gvfs-helper: create tool to fetch objects using the GVFS Protocol

jeffhostetler opened the pull request:
Create helper tool to use the GVFS Protocol REST API to
fetch objects and configuration data from a GVFS cache-server
or a primary Git server.

This is V2.  It replaces my earlier version from
https://github.com/microsoft/git/pull/190

This version refactors pretty much everything and eliminates most
of the code duplication between the various GET and POST forms.
It also clarifies and centralizes the error handling.

This version eliminates the modifications to http.[ch] and credential.[ch]
that I needed in V1.  Everything is now self-contained in gvfs-helper.c.

Usage is the same as before (and described in the TODO comment
at the top of the source file).

jeffhostetler committed:
gvfs-helper: create tool to fetch objects using the GVFS Protocol

Create gvfs-helper.  This is a helper tool to use the GVFS Protocol
REST API to fetch objects and configuration data from a GVFS cache-server
or Git server.  This tool uses libcurl to send object requests to either
server.  This tool creates loose objects and/or packfiles.

Create gvfs-helper-client.  This code resides within git proper and
uses the sub-process API to manage gvfs-helper as a long-running background
process.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

derrickstolee commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2074 @@
+{
+	struct gh__request_params *params = clientp;
+
+//	fprintf(stderr, "TODO [dlt %lld][dln %lld] [ult %lld][uln %lld]\n",
```
(drop commented-out code)

derrickstolee commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2074 @@
+ * later rename it to a proper .pack and run "git index-pack" on
+ * it to create the corresponding .idx file.
+ *
+ * TODO I would rather to just stream the packfile directly into
```
This is an important TODO to handle in a follow-up. Let's create an issue for it.

derrickstolee commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2074 @@
+			  (long)tv.tv_usec);
+	}
+
+	// TODO should this be in the "<ODB>/pack/tempPacks/"
```
Yes, the `tempPacks` dir would more closely match the behavior of `scalar prefetch --folders` _and_ we would avoid issues with hardware failure before a full flush during indexing.

derrickstolee commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2074 @@
+	// TODO consider capturing stdout from index-pack because
+	// TODO it will contain the SHA of the packfile and we can
+	// TODO (should?) add it to the .pack and .idx pathnames
+	// TODO when we install them.
```
It would be best to add it to the names, if possible.

derrickstolee replied:
Keep in mind that it's the last 20 bytes of the .pack, but the last 21-40 bytes of the .idx file.

derrickstolee reviewed (commented):
This is really coming together!

My only big comment is that we'll want to spin out many of the TODOs as issues so we remember to follow up later.

jeffhostetler committed:
sub-process: do not borrow cmd pointer from caller

Teach subprocess_start() to use a copy of the passed `cmd` string
rather than borrowing the buffer from the caller.

Some callers of subprocess_start() pass the value returned from
find_hook() which points to a static buffer and therefore is only
good until the next call to find_hook().  This could cause problems
for the long-running background processes managed by sub-process.c
where later calls to subprocess_find_entry() to get an existing
process will fail.  This could cause more than 1 long-running
process to be created.

TODO Need to confirm, but if only read_object_hook() uses
TODO subprocess_start() in this manner, we could drop this
TODO commit when we drop support for read_object_hook().

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler committed:
sub-process: add subprocess_start_argv()

Add function to start a subprocess with an argv.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

wilbaker commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2424 @@
+	else if (!strcmp(k, "core.gvfs"))
+		gh__cmd_opts.is_core_vfs = 1;
+
+	else if (!strcmp(k, "core.scalar"))
```
Currently Scalar is setting `core.gvfs` (with the specific flags it wants) and not `core.scalar` is that going to cause any issues?

jeffhostetler replied:
No.  I was anticipating that we'd get a core.scalar flag as part of the global rename.  If that's not the case, I can get rid of this.

wilbaker reviewed (commented):
Posting comments/questions that I have so far.  Still looking through the rest of `gvfs-helper.c`

wilbaker commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2424 @@
+/*
+ * Lookup the creds for the main/origin Git server.
+ *
+ * Both VFS and Scalar always require authentication for both
```
FYI - VFS4G and Scalar [do support anonymous access](https://github.com/microsoft/VFSForGit/blob/339b6a62b43501dee21f3dd491055e4bfeecdd92/GVFS/GVFS.Common/Git/GitAuthentication.cs#L173) with the main Git server.  The test repo used for functional tests (https://gvfs.visualstudio.com/ci/_git/ForTests) is configured to allow anonymous access.

It is still the case that the cache-servers always require authentication.

(I don't know this code well enough yet to know if that matters 😄 )

jeffhostetler replied:
I just pushed a fix for this.  Thanks!

jeffhostetler commented on builtin/checkout.c:
```diff
@@ -1146,6 +1147,13 @@ static int switch_branches(const struct checkout_opts *opts,
+				 new_branch_info->commit,
+				 opts->show_progress);
+	exit(1);
+#endif
```
this is still a WIP.  not sure if i need it or not going forward.

jeffhostetler commented on list-objects-filter.c:
```diff
@@ -478,6 +478,111 @@ static void *filter_sparse_oid__init(
+	struct object *obj,
+	const char *pathname,
+	const char *filename,
+	void *filter_data_)
```
There's no official significance to it.
I just did it to try to keep track of the void* callback context handle
and give it a little more meaning than "data" or "cb" or "ctx" as is
usual.

The callback API wants a void *, but the first thing I want to do inside
the function is to cast it to a specific type.  So there's always a dance
in these types of functions to assign void* to a typed alias. This was
my attempt to make the CB API contract a little more meaningful.

jeffhostetler commented on gvfs-helper.c:
```diff
@@ -0,0 +1,2424 @@
+	GH__ERROR_CODE__USAGE = -1, /* will be mapped to usage() */
+	GH__ERROR_CODE__OK = 0,
+	GH__ERROR_CODE__ERROR = 1, /* unspecified */
+//	GH__ERROR_CODE__CACHE_SERVER_NOT_FOUND = 2,
```
I overhauled the error return codes and eliminated the need for this one.  I intend to delete it.

jeffhostetler commented:
Rebased forward to tip of feature branch.  This version currently omits the "--filter=sparse:cone" code because of a massive conflict with upstream.   (And I'm not convinced I still need it.)

jeffhostetler committed:
sha1-file: add function to update existing loose object cache

Create a function to add a new object to the loose object cache
after the existing odb/xx/ directory was scanned.

This will be used in a later commit to keep the loose object
cache fresh after dynamically fetching an individual object
and without requiring the odb/xx/ directory to be rescanned.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler committed:
index-pack: avoid immediate object fetch while parsing packfile

Prevent packfile parsing from accidentally dynamically fetching
each individual object found in the packfile.

When index-pack parses the input packfile, it does a lookup in the
ODB to test for conflicts/collisions.  This can accidentally cause
the object to be individually fetched when gvfs-helper (or
read-object-hook or partial-clone) is enabled.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler committed:
packfile: add install_packed_git_and_mru()

Create a function to install a new packfile into the packed-git
list and add it to the head of the MRU list.

This function will be used in a later commit to install packfiles
created by dynamic object fetching.

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

derrickstolee commented on gvfs-helper-client.c:
```diff
@@ -0,0 +1,378 @@
+		v2_path++;
+
+	prepare_alt_odb(the_repository);
+	for (odb = the_repository->objects->odb; odb; odb = odb->next)
```
@jeffhostetler I think this API could be done significantly differently and make our lives less error prone. The client should pick an ODB based on the config value (and do a platform path conversion to get into some “canonical” form) and then start the helper with that path in mind. The helper could just return a loose object oid or a pack name. We wouldn’t need this loop on every response. We could navigate directly to the odb pointer we have already. 

That would at least help the windows bugs I’m having. The Mac test failure needs some more investigation.

jeffhostetler replied:
I like that.  Let me take a look at that.  That will simplify things considerably.

derrickstolee commented:
@jeffhostetler with your latest changes (3446dd0) I ran the tests with this one failure in the parallel case:

```
System.AggregateException : One or more errors occurred. (  fatal: missing blob object '9d40b0f1651f323765cb38b3e7afe42fab3f0aef'
  Expected: 0
  But was:  128
```

It is being output in this spot of `builtin/rev-list.c`:

```
static inline void finish_object__ma(struct object *obj)
{
	/*
	 * Whether or not we try to dynamically fetch missing objects
	 * from the server, we currently DO NOT have the object.  We
	 * can either print, allow (ignore), or conditionally allow
	 * (ignore) them.
	 */
	switch (arg_missing_action) {
	case MA_ERROR:
		die("missing %s object '%s'",
		    type_name(obj->type), oid_to_hex(&obj->oid));
		return;

	case MA_ALLOW_ANY:
		return;

	case MA_PRINT:
		oidset_insert(&missing_objects, &obj->oid);
		return;

	case MA_ALLOW_PROMISOR:
		if (is_promisor_object(&obj->oid))
			return;
		die("unexpected missing %s object '%s'",
		    type_name(obj->type), oid_to_hex(&obj->oid));
		return;

	default:
		BUG("unhandled missing_action");
		return;
	}
}
```

derrickstolee commented:
I added the following line to `gvfs-helper-client.c` where you already had a `trace2_printf()`:

```
	trace2_data_string("gvfs-helper", the_repository, "ghc__get_immediate", oid_to_hex(oid));
```

This allowed me to run the functional tests with GIT_TRACE2_EVENT pointing to a log file. The relevant events are here:

```

{"event":"cmd_name","sid":"20191002T184920.003088Z-H9764d87b-P00004723","thread":"main","time":"2019-10-02T18:49:20.005234Z","file":"git.c","line":508,"name":"rev-list","hierarchy":"rev-list"}
{"event":"cmd_name","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.005250Z","file":"git.c","line":508,"name":"rev-list","hierarchy":"rev-list"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.011331Z","file":"gvfs-helper-client.c","line":402,"repo":1,"t_abs":0.008517,"t_rel":0.008517,"nesting":1,"category":"gvfs-helper","key":"ghc__get_immediate","value":"9d40b0f1651f323765cb38b3e7afe42fab3f0aef"}
{"event":"region_enter","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.011360Z","file":"gvfs-helper-client.c","line":280,"repo":1,"nesting":1,"category":"gh-client","label":"get"}
{"event":"data","sid":"20191002T184920.003088Z-H9764d87b-P00004723","thread":"main","time":"2019-10-02T18:49:20.011369Z","file":"gvfs-helper-client.c","line":402,"repo":1,"t_abs":0.008553,"t_rel":0.008553,"nesting":1,"category":"gvfs-helper","key":"ghc__get_immediate","value":"9d40b0f1651f323765cb38b3e7afe42fab3f0aef"}
{"event":"region_enter","sid":"20191002T184920.003088Z-H9764d87b-P00004723","thread":"main","time":"2019-10-02T18:49:20.011395Z","file":"gvfs-helper-client.c","line":280,"repo":1,"nesting":1,"category":"gh-client","label":"get"}
{"event":"child_start","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.011410Z","file":"run-command.c","line":736,"child_id":0,"child_class":"subprocess","use_shell":true,"argv":["git","gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"child_start","sid":"20191002T184920.003088Z-H9764d87b-P00004723","thread":"main","time":"2019-10-02T18:49:20.011440Z","file":"run-command.c","line":736,"child_id":0,"child_class":"subprocess","use_shell":true,"argv":["git","gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"version","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724","thread":"main","time":"2019-10-02T18:49:20.019722Z","file":"common-main.c","line":42,"evt":"1","exe":"2.23.0.vfs.1.1.67.g7c2ff67280.dirty"}
{"event":"start","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724","thread":"main","time":"2019-10-02T18:49:20.019762Z","file":"common-main.c","line":43,"t_abs":0.001057,"argv":["/usr/local/libexec/git-core/git","gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"version","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725","thread":"main","time":"2019-10-02T18:49:20.019943Z","file":"common-main.c","line":42,"evt":"1","exe":"2.23.0.vfs.1.1.67.g7c2ff67280.dirty"}
{"event":"start","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725","thread":"main","time":"2019-10-02T18:49:20.019981Z","file":"common-main.c","line":43,"t_abs":0.001033,"argv":["/usr/local/libexec/git-core/git","gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"cmd_name","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724","thread":"main","time":"2019-10-02T18:49:20.020054Z","file":"git.c","line":767,"name":"_run_dashed_","hierarchy":"rev-list/_run_dashed_"}
{"event":"child_start","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724","thread":"main","time":"2019-10-02T18:49:20.020081Z","file":"run-command.c","line":736,"child_id":0,"child_class":"dashed","use_shell":false,"argv":["git-gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"cmd_name","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725","thread":"main","time":"2019-10-02T18:49:20.020241Z","file":"git.c","line":767,"name":"_run_dashed_","hierarchy":"rev-list/_run_dashed_"}
{"event":"child_start","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725","thread":"main","time":"2019-10-02T18:49:20.020270Z","file":"run-command.c","line":736,"child_id":0,"child_class":"dashed","use_shell":false,"argv":["git-gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"version","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034135Z","file":"common-main.c","line":42,"evt":"1","exe":"2.23.0.vfs.1.1.67.g7c2ff67280.dirty"}
{"event":"version","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034160Z","file":"common-main.c","line":42,"evt":"1","exe":"2.23.0.vfs.1.1.67.g7c2ff67280.dirty"}
{"event":"start","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034177Z","file":"common-main.c","line":43,"t_abs":0.000976,"argv":["/usr/local/libexec/git-core/git-gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"start","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034199Z","file":"common-main.c","line":43,"t_abs":0.001049,"argv":["/usr/local/libexec/git-core/git-gvfs-helper","--fallback","--cache-server=trust","--shared-cache=/Users/stolee/Scalar.FT/c295175375894875925597b6e8d5fa37/.customScalarCache/79f28452cdde4b59a77e633508768e1c","server"]}
{"event":"cmd_name","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034220Z","file":"gvfs-helper.c","line":2243,"name":"gvfs-helper","hierarchy":"rev-list/_run_dashed_/gvfs-helper"}
{"event":"cmd_name","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034242Z","file":"gvfs-helper.c","line":2243,"name":"gvfs-helper","hierarchy":"rev-list/_run_dashed_/gvfs-helper"}
{"event":"def_repo","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034388Z","file":"repository.c","line":130,"repo":1,"worktree":"/Users/stolee/Scalar.FT/test/b07cf69385364400a7fd/src"}
{"event":"def_repo","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034406Z","file":"repository.c","line":130,"repo":1,"worktree":"/Users/stolee/Scalar.FT/test/89b74c5133454b7ebb97/src"}
{"event":"cmd_mode","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034574Z","file":"gvfs-helper.c","line":2151,"name":"server"}
{"event":"cmd_mode","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034590Z","file":"gvfs-helper.c","line":2151,"name":"server"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.034663Z","file":"gvfs-helper.c","line":681,"t_abs":0.001475,"t_rel":0.001475,"nesting":1,"category":"gvfs-helper","key":"remote/url","value":"https://gvfs.visualstudio.com/ci/_git/ForTests"}
{"event":"data","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.034678Z","file":"gvfs-helper.c","line":681,"t_abs":0.001535,"t_rel":0.001535,"nesting":1,"category":"gvfs-helper","key":"remote/url","value":"https://gvfs.visualstudio.com/ci/_git/ForTests"}
{"event":"data","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.035705Z","file":"gvfs-helper.c","line":719,"t_abs":0.002558,"t_rel":0.002558,"nesting":1,"category":"gvfs-helper","key":"cache/url","value":"same"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.035747Z","file":"gvfs-helper.c","line":719,"t_abs":0.002512,"t_rel":0.002512,"nesting":1,"category":"gvfs-helper","key":"cache/url","value":"same"}
{"event":"region_enter","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.035818Z","file":"gvfs-helper.c","line":1984,"nesting":1,"category":"gvfs-helper","label":"server/get"}
{"event":"data","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.035835Z","file":"gvfs-helper.c","line":1985,"t_abs":0.002698,"t_rel":0.000021,"nesting":2,"category":"gvfs-helper","key":"server/get/nr_objects","value":"1"}
{"event":"region_enter","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.035854Z","file":"gvfs-helper.c","line":1984,"nesting":1,"category":"gvfs-helper","label":"server/get"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.035878Z","file":"gvfs-helper.c","line":1985,"t_abs":0.002691,"t_rel":0.000035,"nesting":2,"category":"gvfs-helper","key":"server/get/nr_objects","value":"1"}
{"event":"region_leave","sid":"20191002T184920.003083Z-H9764d87b-P00004722/20191002T184920.018944Z-H9764d87b-P00004724/20191002T184920.033411Z-H9764d87b-P00004726","thread":"main","time":"2019-10-02T18:49:20.036156Z","file":"gvfs-helper.c","line":1987,"t_rel":0.000304,"nesting":1,"category":"gvfs-helper","label":"server/get"}
{"event":"region_enter","sid":"20191002T184920.003088Z-H9764d87b-P00004723/20191002T184920.019180Z-H9764d87b-P00004725/20191002T184920.033367Z-H9764d87b-P00004727","thread":"main","time":"2019-10-02T18:49:20.036178Z","file":"gvfs-helper.c","line":565,"nesting":2,"category":"gvfs-helper","label":"GET/objects"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.036209Z","file":"gvfs-helper-client.c","line":347,"repo":1,"t_abs":0.033392,"t_rel":0.024841,"nesting":2,"category":"gh-client","key":"get/count","value":"1"}
{"event":"data","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.036230Z","file":"gvfs-helper-client.c","line":352,"repo":1,"t_abs":0.033422,"t_rel":0.024871,"nesting":2,"category":"gh-client","key":"get/created","value":"nothing"}
{"event":"region_leave","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.036247Z","file":"gvfs-helper-client.c","line":353,"repo":1,"t_rel":0.024889,"nesting":1,"category":"gh-client","label":"get"}
{"event":"error","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.036371Z","file":"usage.c","line":52,"msg":"missing blob object '9d40b0f1651f323765cb38b3e7afe42fab3f0aef'","fmt":"missing %s object '%s'"}
{"event":"exit","sid":"20191002T184920.003083Z-H9764d87b-P00004722","thread":"main","time":"2019-10-02T18:49:20.036399Z","file":"usage.c","line":56,"t_abs":0.033591,"code":128}
```

Specifically, the sid ending in 4722 output the request for the blob _first_, but then was the one to fail. Maybe the blob is being written by the second process as we try to read it in the first process?

derrickstolee commented:
@jeffhostetler I figured out the problem: the two processes are writing to the same temp file at the same time. One "wins" its rename operation and the other loses. The fix would be one of the following:

1. Use a random name, not "oid.temp".
2. Use a lockfile instead of a temp file. If a lock exists, then wait and retry some amount of time.
3. ???

derrickstolee reviewed (approved):
We have functional tests passing. Let's get this in!

jeffhostetler committed:
credential: set trace2_child_class for credential manager children

Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

jeffhostetler commented:
This version is ready for final review/merging.
It includes fixes for the loose tempfile problem.
And it includes streamlined trace2 telemetry.
All of the gvfs-helper code is in 1 large commit at the end of the series.
The earlier commits in the series might be upstreamed even without gvfs-helper.

derrickstolee commented on sub-process.c:
```diff
@@ -78,7 +78,12 @@ int subprocess_start(struct hashmap *hashmap, struct subprocess_entry *entry, co
 	struct child_process *process;
 
-	entry->cmd = cmd;
+	// BUGBUG most callers to subprocess_start() pass in "cmd" the value
```
regarding your TODO in the commit message: we will not be able to drop the read_object_hook() from the Git code as we hope to unify the vfs- and features/sparse-checkout- branches.

jeffhostetler commented:
https://github.com/microsoft/scalar/issues/7
